deadseeker/deadseeker.py

```diff
@@ -0,0 +1,138 @@
+        self.verbose: bool = DEFAULT_VERBOSE
+        self.max_time: int = DEFAULT_RETRY_MAX_TIME
+        self.max_tries: int = DEFAULT_RETRY_MAX_TRIES
+        self.linkacceptor: LinkAcceptor = \
+            LinkAcceptorBuilder()\
+            .addExcludePrefix(*DEFAULT_EXCLUDE_PREFIX).build()
+        self.agent: str = DEFAULT_WEB_AGENT
+
+
+class DeadSeeker:
+    def __init__(self, config: DeadSeekerConfig) -> None:
+        self.config = config
+
+    async def _getUrlFetchResponse(
+            self,
+            session: aiohttp.ClientSession,
+            urlTarget: UrlTarget) -> UrlFetchResponse:
+        start = time.time()  # measure load time (HEAD only)
+        url = urlTarget.url
+        html = ''
+        status = -1
+        end: float = -1
+        message = ''
+        try:
+            async with session.head(
+                    url,
+                    headers={'User-Agent': self.config.agent}) as headResponse:
+                end = time.time()
+                status = headResponse.status
+                if status == 200:
+                    isHtml = 'html' in headResponse.headers['Content-Type']
+                    onSite = urlTarget.home in url
+                    if(isHtml and onSite):
+                        async with session.get(
+                                url,
+                                headers={'User-Agent': self.config.agent}
+                                ) as getResponse:
+                            html = await getResponse.text(encoding="utf-8")
+        except aiohttp.ClientResponseError as e:
+            status = e.status
+        except aiohttp.ClientError as e:
+            message = str(e)
+        if end < 0:
+            end = time.time()
+        elapsed = "{0:.2f} ms".format((end - start)*1000)
+        return UrlFetchResponse(urlTarget, status, elapsed, message, html)
+
+    async def _main(self, urls: List[str]) -> int:
+        num_failures = 0
+        visited: Set[str] = set()
+        targets: Deque[UrlTarget] = Deque()
+        for url in urls:
+            visited.add(url)
+            targets.appendleft(UrlTarget(url, url))
+        linkparser = LinkParser(self.config.linkacceptor)
+        retry_options = ExponentialRetry(
+                            attempts=self.config.max_tries,
+                            exceptions=[aiohttp.ClientError])
+        async with RetryClient(
+                raise_for_status=True, retry_options=retry_options) as session:
+            while targets:
+                tasks = []
+                while targets:
+                    urlTarget = targets.pop()
+                    tasks.append(
+                        asyncio.create_task(
+                            self._getUrlFetchResponse(session, urlTarget)))
+                for task in asyncio.as_completed(tasks):  # completed first
+                    resp = await task
+                    status = resp.status
+                    if(status != 200):
```

Comment:
The status code 200 makes sense, but I think that codes like 204 or 3xx redirects should also be accepted under certain circumstances. Perhaps one could introduce another config variable for this purpose

Reply:
Yeah, that's a really great point, I'll take this out along with some other refactoring that I have been working on

Reply:
@ScholliYT, ready for review!